src/NuGet.Clients/PackageManagement.VisualStudio/ProjectSystems/MSBuildShellOutNuGetProject.cs

```diff
@@ -0,0 +1,405 @@
+            lock (_installedPackagesLock)
+            {
+                _installedPackages = GetPackageReferences(packageSpec);
+            }
+
+            return packageSpec;
+        }
+
+        public override async Task<IEnumerable<PackageReference>> GetInstalledPackagesAsync(CancellationToken token)
+        {
+            await Task.Yield();
+
+            var installedPackages = _installedPackages;
+            if (installedPackages != null)
+            {
+                return installedPackages;
+            }
+
+            lock (_installedPackagesLock)
+            {
+                installedPackages = _installedPackages;
+                if (installedPackages != null)
+                {
+                    return installedPackages;
+                }
+
+                var packageSpec = GetPackageSpecForRestore();
+
+                _installedPackages = GetPackageReferences(packageSpec);
+
+                return _installedPackages;
+            }
+        }
+
+        private List<PackageReference> GetPackageReferences(PackageSpec packageSpec)
+        {
+            if (packageSpec == null)
+            {
+                return new List<PackageReference>();
+            }
+
+            var frameworkSorter = new NuGetFrameworkSorter();
+
+            return packageSpec
+                .TargetFrameworks
+                .SelectMany(f => MapLibrariesToPackages(f.FrameworkName, f.Dependencies))
+                .GroupBy(p => p.PackageIdentity)
+                .Select(g => g.OrderBy(p => p.TargetFramework, frameworkSorter).First())
+                .ToList();
+        }
+
+        public override Task<bool> InstallPackageAsync(
+            PackageIdentity packageIdentity,
+            DownloadResourceResult downloadResourceResult,
+            INuGetProjectContext nuGetProjectContext,
+            CancellationToken token)
+        {
+            throw new NotImplementedException();
+        }
+
+        public override Task<bool> UninstallPackageAsync(
+            PackageIdentity packageIdentity,
+            INuGetProjectContext nuGetProjectContext,
+            CancellationToken token)
+        {
+            throw new NotImplementedException();
+        }
+
+        private IEnumerable<PackageReference> MapLibrariesToPackages(NuGetFramework targetFramework, IEnumerable<LibraryDependency> libraries)
+        {
+            foreach (var library in libraries)
+            {
+                if (library.LibraryRange.TypeConstraint != LibraryDependencyTarget.Package)
+                {
+                    continue;
+                }
+
+                var identity = new PackageIdentity(
+                    library.LibraryRange.Name,
+                    library.LibraryRange.VersionRange.MinVersion);
+
+                yield return new PackageReference(identity, targetFramework);
+            }
+        }
+
+        private static class MsBuildUtility
```

Comment:
Won't fix since this project type will go away soon. I'd rather not have dependencies that would need to sorted out at the point of removing this hack project type.